src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,378 @@
+        /// </summary>
+        private bool _resharperExtensionEnabled = false;
+
+        private bool _infoBarOpen = false;
+
+        [ImportingConstructor]
+        public KeybindingResetDetector(VisualStudioWorkspace workspace, SVsServiceProvider serviceProvider)
+        {
+            _workspace = workspace;
+            _serviceProvider = serviceProvider;
+        }
+
+        public Task InitializeAsync()
+        {
+            // Immediately bail if the user has asked to never see this bar again.
+            if (_workspace.Options.GetOption(KeybindingResetOptions.NeverShowAgain))
+            {
+                return Task.CompletedTask;
+            }
+
+            return InvokeBelowInputPriority(InitializeCore);
+        }
+
+        private void InitializeCore()
+        {
+            AssertIsForeground();
+
+            // Ensure one of the flights is enabled, otherwise bail
+            _experimentationService = _workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!_experimentationService.IsExperimentEnabled(ExternalFlightName) && !_experimentationService.IsExperimentEnabled(InternalFlightName))
+            {
+                return;
+            }
+
+            var vsShell = _serviceProvider.GetService<IVsShell, SVsShell>();
+            var hr = vsShell.IsPackageInstalled(ReSharperPackageGuid, out int extensionEnabled);
+            if (ErrorHandler.Failed(hr))
+            {
+                FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
+                return;
+            }
+
+            _resharperExtensionEnabled = extensionEnabled != 0;
+
+            if (_resharperExtensionEnabled)
+            {
+                // We need to monitor for suspend/resume commands, so create and install the command target and the modal callback.
+                var priorityCommandTargetRegistrar = _serviceProvider.GetService<IVsRegisterPriorityCommandTarget, SVsRegisterPriorityCommandTarget>();
+                hr = priorityCommandTargetRegistrar.RegisterPriorityCommandTarget(
+                    dwReserved: 0 /* from docs must be 0 */,
+                    pCmdTrgt: this,
+                    pdwCookie: out _priorityCommandTargetCookie);
+
+                if (ErrorHandler.Failed(hr))
+                {
+                    FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
+                    return;
+                }
+
+                // Initialize the OleComponent to listen for modal changes (which will tell us when Tools->Options is closed)
+                _oleComponent = OleComponent.CreateHostedComponent("Keybinding Reset Detector");
+                _oleComponent.ModalStateChanged += OnModalStateChanged;
+            }
+
+            UpdateStateMachine();
+        }
+
+        private void UpdateStateMachine()
+        {
+            AssertIsForeground();
+
+            var isEnabled = IsReSharperEnabled();
```

Comment:
Why not just have this function return a ReSharperStatus enum? Then you can just set it once rather than having each branch handle it?

Reply:
Then I think that also convers having both _resharperExtensionEnable and isEnabled that is still confusing.

Reply:
Yeah, then you can delete about a third of the logic in the switch.